pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/ForwardIndexHandler.java

```diff
@@ -677,4 +836,168 @@ private void updateMetadataProperties(File indexDir, Map<String, String> metadat
+
+    // Remove range index, inverted index and FST index.
+    removeDictRelatedIndexes(column, segmentWriter);
+
+    // Delete marker file.
+    FileUtils.deleteQuietly(inProgress);
+
+    LOGGER.info("Created raw based forward index for segment: {}, column: {}", segmentName, column);
+  }
+
+  private void rewriteDictToRawForwardIndex(String column, ColumnMetadata existingColMetadata,
+      SegmentDirectory.Writer segmentWriter, File indexDir, IndexCreatorProvider indexCreatorProvider)
+      throws Exception {
+    try (ForwardIndexReader reader = LoaderUtils.getForwardIndexReader(segmentWriter, existingColMetadata)) {
+      Dictionary dictionary = LoaderUtils.getDictionary(segmentWriter, existingColMetadata);
+      // lengthOfLongestEntry not available for dict columns. Should be computed by reading all values in dictionary.
+      IndexCreationContext.Builder builder =
+          IndexCreationContext.builder().withIndexDir(indexDir).withColumnMetadata(existingColMetadata);
+      builder.withDictionary(false);
+
+      if (existingColMetadata.isSingleValue()) {
+        int lengthOfLongestEntry = getLengthOfLongestEntryFromDictionary(column, reader, dictionary);
+        builder.withLengthOfLongestEntry(lengthOfLongestEntry);
+      } else {
+        FieldSpec.DataType dataType = existingColMetadata.getDataType();
+        boolean isFixedWidth = dataType.getStoredType().isFixedWidth();
+
+        if (!isFixedWidth) {
+          // For variable length stored types, maxRowLengthInBytes is required to create the forwardIndexCreator.
+          // This can only be determined by reading the entire MV forward index.
+          int maxRowLength = getMaxRowLengthForMVColumn(column, reader, dictionary);
+          builder.withMaxRowLengthInBytes(maxRowLength);
+        }
+      }
+
+      Map<String, ChunkCompressionType> compressionConfigs = _indexLoadingConfig.getCompressionConfigs();
+      ChunkCompressionType compressionType;
+      if (compressionConfigs.containsKey(column)) {
+        compressionType = compressionConfigs.get(column);
+      } else {
+        compressionType = SegmentColumnarIndexCreator.getDefaultCompressionType(existingColMetadata.getFieldType());
+      }
+
+      IndexCreationContext.Forward context =
+          builder.build().forForwardIndex(compressionType, _indexLoadingConfig.getColumnProperties());
+      try (ForwardIndexCreator creator = indexCreatorProvider.newForwardIndexCreator(context)) {
+        int numDocs = existingColMetadata.getTotalDocs();
+        forwardIndexRewriteHelper(column, existingColMetadata, reader, creator, numDocs, null, dictionary);
+      }
+    }
+  }
+
+  private int getMaxRowLengthForMVColumn(String column, ForwardIndexReader reader, Dictionary dictionary)
+      throws Exception {
+    ColumnMetadata existingColMetadata = _segmentMetadata.getColumnMetadataFor(column);
+    AbstractColumnStatisticsCollector statsCollector =
```

Comment:
Probably nothing wrong in theory but I wonder why do we need to go through StatsCollector framework to gather this info. We can simply read the forward index to collect this right ?

Reply:
IMO, using statsCollector is convenient and it can be the single place that requires change if we alter anything in the future. Without StatsCollector, we might have duplicate the logic for each datatype here that might lead to unnecessarily more code?

Reply:
I see. Fair enough